### PR TITLE
Modifying decoders and attention for vllm

### PR DESCRIPTION
**Pull request by NicoGrande**
# Description

This PR introduces new optional arguments `kv_cache` and `attention_metadata` for model decoder blocks and for the `attentions.py` module. These arguments are provided by vLLM when executing a MaxText model from vLLM Engine and are used when calling the ragged paged attention kernel in `tpu-inference`.

This PR builds on the work started in #2612.

**Note: This PR changes the expected method signature from Attention layers and decoder layers.**

# Tests

Includes a new unit-test in `attention_test.py`. 

Additionally, end-to-end tests were performed locally on a v6e VM using the following test command:

In one process, start the vLLM server. This requires a local `config.json` file for the corresponding model you are trying to test from HuggingFace. Modify this file such that `architectures: "MaxTextForCausalLM"` is set.
```
HF_TOKEN=<YOUR_HF_TOKEN> TPU_BACKEND_TYPE=jax \
  python -m vllm.entrypoints.cli.main serve \
  <HF_MODEL> \
  --max-num-batched-tokens=32 \
  --max-model-len=32 \
  --max-num-seqs=1 \
  --tensor-parallel-size=4 \
  --hf_config_path=<PATH_TO_LOCAL_HF_CONFIG_JSON> \
  --additional-config='{"maxtext_config": {"model_name": "<MAXTEXT_MODEL_NAME>", "max_prefill_predict_length": 28, "max_target_length": 32, "ici_tensor_parallelism": 4, "load_parameters_path": "<MAXTEXT_CHECKPOINT_PATH>"}}'
```

In a second process, issue the query to the model:
```
curl http://localhost:8000/v1/completions \
   -H "Content-Type: application/json" \
   -d '{
       "model": "<HF_MODEL>",
       "prompt": ["Seattle is a"],
       "max_tokens": 16,
       "temperature": 0
   }'
```
Results for different tested models are shown below:
## llama3.1-8b
```
output: " city that is known for its coffee culture, and it's not hard to see"
```

## gemma3-4b
```
output: " vibrant city with a lot to offer, and it's a great place to"
```

## qwen3-8b
```
output: " city in the state of Washington, in the Pacific Northwest region of the United States"
```
# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/layers/models.py [hunk 1]
```diff
@@ -386,6 +397,8 @@ def __call__(
       decoder_target_tokens: Target tokens for the decoder (optional, used in MTP).
       decoder_target_mask: Target mask for the decoder (optional, used in MTP).
       nnx_method: Method to call on the NNX module (optional).
+      kv_caches: List of KV caches for each attention layer, used when invoking from vLLM (optional).
+      attention_metadata: Mapping to store attention metadata, used when invoking from vLLM (optional).
 
     Returns:
       Logits from the Transformer model.
```
Comment (gagika) [inline src/MaxText/layers/models.py]: Could you update return to be `logits, hidden_state, kv_caches` if called from vLLM

File: src/MaxText/layers/attentions.py [hunk 2]
```diff
@@ -1000,6 +1040,15 @@ def __call__(
           query, key, value, decoder_segment_ids, model_mode, previous_chunk, slot=slot, page_state=page_state
       )
       out = unnormalized_out / (exp_sum + 1e-9) if exp_sum is not None else unnormalized_out
+
+    elif self.config.attention == "vllm_rpa" and attention_metadata and model_mode != MODEL_MODE_TRAIN:
+      batch, seq_len, num_heads, head_dim = query.shape
```
Comment (cgarciae) [inline src/MaxText/layers/attentions.py:1080]: if `self.config.attention == "vllm_rpa"` but there is not attention metadata should we raise instead?

Comment (NicoGrande) [reply src/MaxText/layers/attentions.py:1080]: Great catch! This is actually a small hack I put together to pass the MaxText model compilation stage. To get around this, the more correct approach would be to pass a `dummy_attention_metadata` object in the following line: https://github.com/AI-Hypercomputer/maxtext/blob/main/src/MaxText/layers/models.py#L307-L311 

If we do that, then we can raise an error inside the `elif` to fail in case no `attention_metadata` is provided.

File: src/MaxText/layers/models.py [hunk 3]
```diff
@@ -199,6 +204,10 @@ def __call__(
           model_mode=model_mode,
       )
 
+    if self.config.attention == "vllm_rpa":
+      # In vLLM, logits are computed separately after updating the KV cache.
+      return logits, hidden_state, kv_caches
```
Comment (gagika) [inline src/MaxText/layers/models.py:211]: We little bit discussed it in person

Can we skip computing the logits for vllm_rpa use case as it might not be needed (and when it's needed it can be computed from hidden_state) and this will avoid need for caching it?

it's not urgent, can be a follow up PR.

**Review by gagika — APPROVED** [review]
Thanks Nico,
Feel free to address the comments in a follow up PR.

File: src/MaxText/layers/models.py (hunk 1 above)
Comment (NicoGrande) [reply src/MaxText/layers/models.py]: Great, added this!

File: src/MaxText/layers/attentions.py (hunk 2 above)
Comment (NicoGrande) [reply src/MaxText/layers/attentions.py:1080]: Done!

File: src/MaxText/layers/models.py (hunk 3 above)
Comment (NicoGrande) [reply src/MaxText/layers/models.py:211]: Sounds great, will add this in a follow up PR.

File: src/MaxText/layers/attention_op.py [hunk 4]
```diff
@@ -852,6 +852,7 @@ def apply_attention(
         or (self.attention_kernel == "autoselected" and model_mode == MODEL_MODE_AUTOREGRESSIVE)
         or (self.attention_kernel == "autoselected" and length < 128)
         or (self.attention_kernel == "paged")
+        or (self.attention_kernel == "vllm_rpa")
```
Comment (gagika) [inline src/MaxText/layers/attention_op.py:857]: could you comment somewhere:

```
# 'vllm_rpa' uses the same dot-attention wrapper but routes to the vLLM
# ragged paged attention kernel in `Attention.__call__`.
```

**Review by gagika — APPROVED** [review]

File: src/MaxText/layers/attention_op.py (hunk 4 above)
Comment (NicoGrande) [reply src/MaxText/layers/attention_op.py:857]: Added!
